src/IO/H5/Helpers.cpp

```diff
@@ -339,24 +442,63 @@ bool contains_attribute(const hid_t file_id, const std::string& group_name,
          std::end(names);
 }
 
-DataVector read_data(const hid_t group_id, const std::string& dataset_name) {
-  // Read a DataVector from the group
-  const hid_t dataset_id =
-      H5Dopen2(group_id, dataset_name.c_str(), h5p_default());
-  CHECK_H5(dataset_id, "could not open dataset '" << dataset_name << "'");
-  // Get the number of points. These do not need a "close" call.
-  const hid_t space_id = H5Dget_space(dataset_id);
-  CHECK_H5(space_id, "Failed to open dataspace");
-  const hid_t number_of_points = H5Sget_simple_extent_npoints(space_id);
-  CHECK_H5(number_of_points, "Failed to get number of points");
-  H5Sclose(space_id);
-  // Load the data
-  DataVector data(static_cast<size_t>(number_of_points));
-  CHECK_H5(H5Dread(dataset_id, h5_type<double>(), h5s_all(), h5s_all(),
-                   h5p_default(), static_cast<void*>(data.data())),
-           "Failed to read data");
+hid_t open_dataset(const hid_t group_id,
+                   const std::string& dataset_name) noexcept {
+  const hid_t dataset_id = H5Dopen2(
+                           group_id, dataset_name.c_str(), h5p_default());
+  CHECK_H5(dataset_id, "Failed to open dataset '" << dataset_name << "'");
+  return dataset_id;
+}
+
+void close_dataset(const hid_t dataset_id) noexcept {
   CHECK_H5(H5Dclose(dataset_id), "Failed to close dataset");
-  return data;
+}
+
+hid_t open_dataspace(const hid_t dataset_id) noexcept {
+  const hid_t dataspace_id = H5Dget_space(dataset_id);
+  CHECK_H5(dataspace_id, "Failed to open dataspace");
+  return dataspace_id;
+}
+
+void close_dataspace(const hid_t dataspace_id) noexcept {
+  CHECK_H5(H5Sclose(dataspace_id), "Failed to close dataspace");
+}
+
+template <size_t Rank, typename T>
+T read_data(const hid_t group_id, const std::string& dataset_name) noexcept {
+  const hid_t dataset_id = open_dataset(group_id, dataset_name);
+  const hid_t dataspace_id = open_dataspace(dataset_id);
+  std::array<hsize_t, Rank> size{}, max_size{};
+  if (static_cast<int>(Rank) !=
+      H5Sget_simple_extent_dims(dataspace_id, size.data(), max_size.data())) {
```

Comment:
If it was undefined, wouldn't it trigger the `ERROR`?

Reply:
No, that's the trouble with UB. Just because it works now doesn't mean it is guaranteed to work and we want our code to be "guaranteed" to work :) Things like turning on optimizations, using a different compiler, or even running tests in a different order could produce wrong results. This is the trouble with UB, it's very difficult to test for. The best way is to read the documentation and see what the requirements and guarantees are. In the case of std::array the pointer returned by `data()` is not dereferenceable, and if HDF5 requires it be, then you have UB. "If size() is 0, data() may or may not return a null pointer." from [here](https://en.cppreference.com/w/cpp/container/array/data)

Reply:
Ok, so I guess I should go back to what I had before and just say `size` is 1 for the `Rank=0` case then?

Reply:
Have you determined what H5Sget_simple_extent_dims does in the case of Rank 0? What would this do if the Rank you passed is 1 but the written rank is 4? Is that UB? It looks like it could be, but I don't know. These are things that need to be understood before a best-solution can be found. I don't know the answers to those questions, but they should be documented in HDF5, and if not one could look at the HDF5 code and figure out what they assume.

Reply:
[here's](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5S.html#Dataspace-ExtentDims) the HDF5 spec

Reply:
I'm not sure I understand what you mean by the question "What would this do if the Rank you passed is 1 but the written rank is 4?" If the written rank is 4, H5Sget_simple_extent_dims would return 4, in which case, if I pass 'Rank=1', the if statement is not satisfied and the `ERROR` is triggered. Maybe that's not what you meant?

It's true that I'm not sure that H5Sget_simple_extent_dims will always return 0 if the size of the array given to it is 0. If it does then I guess there's no problem but I do need to check that I suppose.

Reply:
Ok, thanks, this works!